### PR TITLE
S-57 → S-101: complex/simple attribute mapping improvements (Refs #433)

### PR DESCRIPTION
**Pull request by philliphoff**
## Summary

Audit-driven improvements to the S-57 → S-101 transformer (`EncDotNet.S100.Datasets.S57`), validated against a 3,636-cell NOAA ENC corpus. Each change closes a real mapping gap for S-57 attributes that had no S-101 home, is gated on the S-101 Feature Catalogue's per-feature bindings, and carries xunit coverage. Refs #433.

## Attribute mappings added

| S-57 | → S-101 | Notes |
|---|---|---|
| `DATSTA`/`DATEND`, `PERSTA`/`PEREND`, `SURSTA`/`SUREND` | `fixedDateRange` / `periodicDateRange` / `surveyDateRange` (`dateStart`/`dateEnd`) | Three date-range complexes; consumer scope-delimiter fix |
| `CATZOC` | `zoneOfConfidence` (`categoryOfZoneOfConfidenceInData`) | |
| `CATPRA` | production-area split (feature-dependent) | |
| `NATSUR`/`NATQUA` | `surfaceCharacteristics` (`natureOfSurface` + `natureOfSurfaceQualifyingTerms`) | positional list pairing |
| `SIGSEQ` | `signalSequence` (nested in `rhythmOfLight` + top-level on `FogSignal`/`RadarTransponderBeacon`) | |
| sector lights | `LightSectored` / `sectorCharacteristics` / `lightSector` / `sectorLimit` | three-level sector geometry |
| `HORCLR` | `horizontalClearanceOpen` (Gate) / `horizontalClearanceFixed` (spans/tunnels/etc.) | |
| `SORDAT` | `reportedDate` (top-level simple, feature binding-gated) | |
| `VALLMA` | `valueOfLocalMagneticAnomaly` (`magneticAnomalyValue`) | |
| `RADWAL` | `radarWaveLength` (`waveLengthValue` + `radarBand`, one instance per `value-band` pair) | |
| `CURVEL` | `speed` (`speedMaximum`) on `CurrentNonGravitational` / `TidalStreamFloodEbb` | |
| `MLTYLT` | `multiplicityOfFeatures` (`multiplicityKnown` + `numberOfFeatures`) on the light classes | first boolean sub-attribute |

## Validation

Full-corpus audit after each change confirmed:
- Feature emission unchanged (3,666,338 features) — no regressions.
- 0 crashes across all 3,636 cells.
- No new dropped-enum values.
- Targeted attributes driven to 0 unmapped where a conformant S-101 home exists.

Remaining unmapped attributes were each confirmed **by design** via the Feature Catalogue (no conformant per-feature target): `SORIND`, `SORDAT`-residual (non-binding features), `CATCOV`, `HORCLR`/Bridge-residual, `POSACC`, `CATRUN`, `SOUACC`, `NATQUA`-residual, `PILDST`, `VERACC`.

## Tests

232 S-57 translator tests pass (`dotnet test --configuration Release`), plus previously-verified S-101 (101) and pipeline (880) suites remain green. Each mapping has success + non-binding-feature + edge-case coverage.

### CODE REVIEW

File: src/EncDotNet.S100.Datasets.S57/S57ToS101Translator.cs [hunk 1]
```diff
@@ -389,29 +781,210 @@ private IReadOnlyList<S101Attribute> TranslateAttributes(
...
             foreach (var a in attrs)
             {
-                // Textual-info attributes are handled as a complex attribute
-                // group below — skip the per-attribute pass-through.
-                if (a.AttributeCode is S57AttrInform or S57AttrNinfom or S57AttrTxtdsc or S57AttrNtxtds)
+                // Textual-info and object-name attributes are handled as complex
+                // attribute groups below — skip the per-attribute pass-through.
+                if (a.AttributeCode is S57AttrInform or S57AttrNinfom or S57AttrTxtdsc or S57AttrNtxtds
+                    or S57AttrObjnam or S57AttrNobjnm)
+                    continue;
```
[the remark this answers is not in the document]
Comment (philliphoff) [reply src/EncDotNet.S100.Datasets.S57/S57ToS101Translator.cs]: Accepted. OBJNAM/NOBJNM are now split out of the unconditional pass-through skip and gated on `_featureBindings.Binds(feature.S101Code, "featureName")`. On a feature that does not bind `featureName` (e.g. QualityOfBathymetricData / M_QUAL, OBJL 308) they now fall through to the no-rule path and are recorded as unmapped, matching the pattern used by all the other assembled complexes. Added a test asserting the diagnostics record the drop; the INFORM/NINFOM/TXTDSC/NTXTDS information handling is unchanged. Fixed in f0ab68f.

File: src/EncDotNet.S100.Datasets.S57/S57ToS101Translator.cs [hunk 2]
```diff
@@ -425,6 +998,186 @@ private IReadOnlyList<S101Attribute> TranslateAttributes(
...
                 AppendInformationInstance(builder, text: ninfomText, fileReference: ntxtdsFile, language: string.Empty);
 
+            // Append `featureName` complex-attribute instances. OBJNAM carries
+            // the English name; NOBJNM the national-language name (emitted with
+            // an empty language string, as S-57 carries no language tag —
+            // mirrors the INFORM/NINFOM handling above).
+            if (objnamText is not null)
+                AppendFeatureNameInstance(builder, name: objnamText, language: LanguageEng);
+            if (nobjnmText is not null)
+                AppendFeatureNameInstance(builder, name: nobjnmText, language: string.Empty);
```
[the remark this answers is not in the document]
Comment (philliphoff) [reply src/EncDotNet.S100.Datasets.S57/S57ToS101Translator.cs]: Accepted. The `featureName` emission block is now wrapped in `if (bindsFeatureName)`, so no `featureName` instance is appended on feature classes that do not bind it in the FC (e.g. QualityOfBathymetricData). This is the same per-feature binding gate the other assembled complexes use. Verified via the full-corpus audit (feature emission unchanged, 0 crashes) plus a unit test on OBJL 308 asserting no `featureName` instance is emitted. Fixed in f0ab68f.

File: src/EncDotNet.S100.Datasets.S57/S57ToS101Translator.cs [hunk 3]
```diff
@@ -61,13 +63,249 @@ public sealed class S57ToS101Translator
...
+    private const string S101AttrFeatureName = "featureName";
+    private const string S101AttrName = "name";
+
+    // ── S-57 light-characteristic attribute codes (S-57 Appendix A) ──
+    // On light features these do NOT pass through as simple attributes;
+    // they become sub-attributes of the S-101 `rhythmOfLight` complex
+    // attribute. LITCHR is the mandatory `lightCharacteristic` [1..1];
+    // SIGGRP/SIGPER are the optional `signalGroup`/`signalPeriod`. (The
+    // nested `signalSequence` sub-complex from SIGSEQ, and the sector
+    // geometry from SECTR1/SECTR2/LITVIS, are deferred.)
```
[the remark this answers is not in the document]
Comment (philliphoff) [reply src/EncDotNet.S100.Datasets.S57/S57ToS101Translator.cs]: Accepted. Updated the comment: the nested signalSequence sub-complex from SIGSEQ is now assembled, and sector lights (SECTR1/SECTR2) redirect to LightSectored with a sectorCharacteristics complex, so nothing here is deferred anymore.

File: src/EncDotNet.S100.Datasets.S57/S57ToS101Translator.cs [hunk 4]
```diff
@@ -451,6 +1215,414 @@ private void AppendInformationInstance(
...
+            var langCode = GetOrAssignAttributeCode(S101AttrLanguage);
+            builder.Add(new S101Attribute(langCode, 1, language));
+        }
+
+        // Emits a `rhythmOfLight` complex-attribute instance (marker +
+        // lightCharacteristic + optional signalGroup/signalPeriod), using the
+        // same flat marker + contiguous-sub-attribute convention as the other
+        // complex attributes. Only the first level of the FC's structure is
+        // populated; the nested `signalSequence` sub-complex (from SIGSEQ) is
+        // not yet assembled.
```
[the remark this answers is not in the document]
Comment (philliphoff) [reply src/EncDotNet.S100.Datasets.S57/S57ToS101Translator.cs]: Accepted. Reworded the summary so it no longer claims signalSequence is not assembled; it now states that when SIGSEQ supplies a signalSequence it is appended as nested sub-complex instances after signalGroup/signalPeriod.

File: src/EncDotNet.S100.Datasets.S57/S57ToS101Translator.cs [hunk 5]
```diff
@@ -451,6 +1215,414 @@ private void AppendInformationInstance(
...
+                {
+                    builder.Add(new S101Attribute(GetOrAssignAttributeCode(S101AttrSectorLimitOne), 1, string.Empty));
+                    builder.Add(new S101Attribute(GetOrAssignAttributeCode(S101AttrSectorBearing), 1, sectorBearingOne));
+                }
+                if (sectorBearingTwo is not null)
+                {
+                    builder.Add(new S101Attribute(GetOrAssignAttributeCode(S101AttrSectorLimitTwo), 1, string.Empty));
+                    builder.Add(new S101Attribute(GetOrAssignAttributeCode(S101AttrSectorBearing), 1, sectorBearingTwo));
+                }
+            }
```
[the remark this answers is not in the document]
Comment (philliphoff) [reply src/EncDotNet.S100.Datasets.S57/S57ToS101Translator.cs]: Accepted. sectorLimitOne and sectorLimitTwo are both mandatory [1..1] in the FC, so the sectorLimit subtree (itself [0..1]) is now emitted only when both bearings are present; a lone bearing omits the whole subtree. Added a regression test.

File: src/EncDotNet.S100.Datasets.S57/S57ToS101Translator.cs [hunk 6]
```diff
@@ -61,13 +63,249 @@ public sealed class S57ToS101Translator
...
+    private const string S101AttrSignalPeriod = "signalPeriod";
+
+    // S-101 feature classes that bind `rhythmOfLight` directly (per the
+    // bundled S-101 FC). On any other feature class, LITCHR/SIGGRP/SIGPER
+    // are handled by the normal per-attribute path (e.g. `signalGroup` /
+    // `signalPeriod` are directly feature-bound simple attributes on
+    // FogSignal / RadarTransponderBeacon).
+    private static readonly FrozenSet<string> RhythmOfLightFeatureClasses =
+        new[] { "LightAllAround", "LightFogDetector", "LightAirObstruction" }
+            .ToFrozenSet(StringComparer.Ordinal);
```
[the remark this answers is not in the document]
Comment (philliphoff) [reply src/EncDotNet.S100.Datasets.S57/S57ToS101Translator.cs]: Accepted. Replaced the hard-coded FrozenSet with _featureBindings.Binds(code, rhythmOfLight) so bindings come from one FC-derived source, and removed the System.Collections.Frozen import. Verified the FC yields the identical set {LightAllAround, LightFogDetector, LightAirObstruction}, and the full-corpus audit shows unchanged emission.

File: src/EncDotNet.S100.Datasets.S57/S57ToS101Translator.cs [hunk 7]
```diff
@@ -451,6 +1205,412 @@ private void AppendInformationInstance(
...
+            foreach (var colour in SplitEnumList(colourList))
+            {
+                if (_allowedEnumValues is not null
+                    && !_allowedEnumValues.IsAllowed(S101AttrColour, colour))
+                {
+                    _diagnostics?.RecordDroppedEnumValue(S101AttrColour, colour);
+                    continue;
+                }
+                builder.Add(new S101Attribute(GetOrAssignAttributeCode(S101AttrColour), 1, colour));
+            }
```
[the remark this answers is not in the document]
Comment (philliphoff) [reply src/EncDotNet.S100.Datasets.S57/S57ToS101Translator.cs:1651]: Accepted. AppendSectorCharacteristicsInstance now captures the builder start index and rolls back the whole sectorCharacteristics instance when no valid colour row is produced (COLOUR missing or every code FC-rejected), so no partial lightSector is emitted; the drop is recorded against COLOUR in diagnostics. Added a regression test, and full-corpus feature emission is unchanged.

File: tests/EncDotNet.S100.Datasets.S57.Tests/S57TranslationDiagnosticsTests.cs [hunk 8]
```diff
@@ -0,0 +1,233 @@
+using EncDotNet.S57;
+
+namespace EncDotNet.S100.Datasets.S57.Tests;
```
[the remark this answers is not in the document]
Comment (philliphoff) [reply tests/EncDotNet.S100.Datasets.S57.Tests/S57TranslationDiagnosticsTests.cs]: Accepted and removed the unused `using EncDotNet.S57;`. All S57 types in this test file are already fully qualified, so it was genuinely unused. (Note: CS8019 is not currently enabled-as-error in our build config, so this was not actually breaking CI, but the cleanup is correct.)


_Auto-resolved by the [GitHub Copilot app](https://github.com/github/app)._

File: src/EncDotNet.S100.Datasets.S57/S57S101Mapping.cs [hunk 9]
```diff
@@ -98,7 +98,9 @@ private S57S101Mapping(
         foreach (var redirect in rule.Redirects)
         {
             if (s57AttributesByAcronym.TryGetValue(redirect.ConditionAttribute, out var v)
-                && redirect.ConditionValues.Contains(v, StringComparer.Ordinal))
+                && (redirect.ConditionPresent
+                    ? !string.IsNullOrEmpty(v)
+                    : redirect.ConditionValues.Contains(v, StringComparer.Ordinal)))
```
[the remark this answers is not in the document]
Comment (philliphoff) [reply src/EncDotNet.S100.Datasets.S57/S57S101Mapping.cs:123]: Accepted. A value-matching redirect (ConditionPresent == false) with no ConditionValues can never match, so this now fails fast with a clear ArgumentException at mapping construction (the single choke point shared by BuildDefault and Builder), rather than per-feature in the ResolveFeature hot path. Added two tests covering the throw and the valid presence-only redirect.


_Auto-resolved by the [GitHub Copilot app](https://github.com/github/app)._

File: src/EncDotNet.S100.Datasets.S101/S101LuaDataProvider.cs [hunk 10]
```diff
@@ -632,6 +735,9 @@ private IReadOnlyList<S101Attribute> ResolveAttributeScope(IReadOnlyList<S101Att
 
             if (numericCode is null) return [];
 
+            var complexCodes = GetComplexAttributeNumericCodes();
+            var nestedDescendants = GetNestedDescendantComplexCodes(numericCode.Value);
+
```
[the remark this answers is not in the document]
Comment (philliphoff) [reply src/EncDotNet.S100.Datasets.S101/S101LuaDataProvider.cs:750]: Accepted. ResolveAttributeScope now validates the colon position and uses int.TryParse, returning an empty scope for any malformed path (missing colon, empty code, empty or non-numeric index) instead of throwing and crashing the portrayal pipeline. Added a parameterized test covering all four malformed forms.


_Auto-resolved by the [GitHub Copilot app](https://github.com/github/app)._

File: src/EncDotNet.S100.Datasets.S57/S57ToS101Translator.cs [hunk 11]
```diff
@@ -451,6 +1205,430 @@ private void AppendInformationInstance(
...
+                    _diagnostics?.RecordRuleDroppedAttribute(S57AttrRadwal);
+                    continue;
+                }
+
+                var complexCode = GetOrAssignAttributeCode(S101AttrRadarWaveLength);
+                // Marker entry — Index=1, value=empty — followed by the sub-attributes.
+                builder.Add(new S101Attribute(complexCode, 1, string.Empty));
+                builder.Add(new S101Attribute(GetOrAssignAttributeCode(S101AttrWaveLengthValue), 1, value));
+                builder.Add(new S101Attribute(GetOrAssignAttributeCode(S101AttrRadarBand), 1, band));
+            }
```
[the remark this answers is not in the document]
Comment (philliphoff) [reply src/EncDotNet.S100.Datasets.S57/S57ToS101Translator.cs:1409]: Accepted. AppendRadarWaveLengthInstances now caps output at the FC upper bound of 2 for radarWaveLength; surplus RADWAL pairs are dropped and recorded via diagnostics so the dataset stays conformant. Added a test asserting a three-pair input emits exactly two instances plus a drop.


_Auto-resolved by the [GitHub Copilot app](https://github.com/github/app)._

File: src/EncDotNet.S100.Datasets.S57/S57TranslationDiagnostics.cs [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+/// </summary>
+/// <param name="ObjectClass">S-57 numeric object class (OBJL) that owned the attribute.</param>
+/// <param name="AttributeCode">S-57 numeric attribute code (ATTL) that was dropped.</param>
+public readonly record struct S57AttributeDrop(ushort ObjectClass, ushort AttributeCode);
+
+/// <summary>
+/// Identifies an S-101 enumerate value that was dropped during translation
+/// because the destination S-101 Feature Catalogue does not list it as an
+/// allowable code for the target attribute (see
+/// <see cref="S101AllowedEnumValues"/>).
```
[the remark this answers is not in the document]
Comment (philliphoff) [reply src/EncDotNet.S100.Datasets.S57/S57TranslationDiagnostics.cs]: Accepted. Fixed the typo: "enumerate value" is now "enumerated value".


_Auto-resolved by the [GitHub Copilot app](https://github.com/github/app)._

File: src/EncDotNet.S100.Datasets.S57/S57TranslationDiagnostics.cs [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+
+    /// <summary>
+    /// S-57 attributes (ATTL → occurrence count) that had a rule but resolved to
+    /// nothing — the rule (or a feature-level override) maps the value to
+    /// <c>null</c>, i.e. an intentional value-level drop.
+    /// </summary>
+    public IReadOnlyDictionary<ushort, int> RuleDroppedAttributes => _ruleDroppedAttributes;
+
+    /// <summary>
+    /// S-101 enumerate values dropped because the destination Feature Catalogue
```
[the remark this answers is not in the document]
Comment (philliphoff) [reply src/EncDotNet.S100.Datasets.S57/S57TranslationDiagnostics.cs]: Accepted. Fixed this second occurrence: "enumerate values" is now "enumerated values".


_Auto-resolved by the [GitHub Copilot app](https://github.com/github/app)._

File: src/EncDotNet.S100.Datasets.S57/S57ToS101Translator.cs [hunk 14]
```diff
@@ -61,13 +62,240 @@ public sealed class S57ToS101Translator
...
+    // codes (verified against the bundled FC).
+    private const string S101AttrFixedDateRange = "fixedDateRange";
+    private const string S101AttrPeriodicDateRange = "periodicDateRange";
+    private const string S101AttrSurveyDateRange = "surveyDateRange";
+    private const string S101AttrDateStart = "dateStart";
+    private const string S101AttrDateEnd = "dateEnd";
+
+    // S-57 CATZOC (Category of zone of confidence in data, ATTL 72) maps to the
+    // S-101 `zoneOfConfidence` complex attribute's `categoryOfZoneOfConfidence-
+    // InData` sub-attribute (S-101 Conversion Guidance; verified against the
```
[the remark this answers is not in the document]
Comment (philliphoff) [reply src/EncDotNet.S100.Datasets.S57/S57ToS101Translator.cs]: Accepted. The comment no longer breaks the code across a hyphen+newline; it now reads the correct sub-attribute name categoryOfZoneOfConfidenceInData (matching the S101AttrCategoryOfZocInData constant).


_Auto-resolved by the [GitHub Copilot app](https://github.com/github/app)._

File: src/EncDotNet.S100.Datasets.S57/DefaultRules.cs [hunk 15]
```diff
@@ -136,6 +161,141 @@ public static IEnumerable<S57FeatureRule> FeatureRules()
...
+        // Meta (M_*) objects. In S-101 several S-57 meta objects become
+        // first-class features carrying data-quality / coverage information.
+        // M_QUAL carries CATZOC (a bathymetric-quality concept) and therefore
+        // converts to QualityOfBathymetricData; QualityOfNonBathymetricData is
+        // sourced from M_ACCY (S-57 → S-101 Conversion Guidance).
+        yield return F(302, "M_COVR", "DataCoverage");
+        yield return F(305, "M_NPUB", "InformationArea");
+        yield return F(306, "M_NSYS", "NavigationalSystemOfMarks");
+        yield return F(308, "M_QUAL", "QualityOfBathymetricData");
+        yield return F(309, "M_SDAT", "SoundingDatum");
```
[the remark this answers is not in the document]
Comment (philliphoff) [reply src/EncDotNet.S100.Datasets.S57/DefaultRules.cs:298]: Accepted. This was a genuinely missing rule, not just a comment issue: added F(301, "M_ACCY", "QualityOfNonBathymetricData") so M_ACCY records now translate to the aliased S-101 feature instead of being dropped as unmapped.


_Auto-resolved by the [GitHub Copilot app](https://github.com/github/app)._

File: tests/EncDotNet.S100.Datasets.S57.Tests/S57S101MappingTests.cs [hunk 16]
```diff
@@ -332,4 +374,123 @@ public void NewFeatureClasses_MapToExpectedS101Code(ushort objl, string expected
...
+    {
+        var m = S57S101Mapping.Default;
+        Assert.Equal(expected, m.ResolveFeatureCode(objl));
+    }
+
+    [Theory]
+    [InlineData(302, "DataCoverage")]                // M_COVR
+    [InlineData(305, "InformationArea")]             // M_NPUB
+    [InlineData(306, "NavigationalSystemOfMarks")]   // M_NSYS
+    [InlineData(308, "QualityOfBathymetricData")]    // M_QUAL
```
[the remark this answers is not in the document]
Comment (philliphoff) [reply tests/EncDotNet.S100.Datasets.S57.Tests/S57S101MappingTests.cs:410]: Accepted. Added [InlineData(301, "QualityOfNonBathymetricData")] // M_ACCY to GapMetaObjects_MapToExpectedS101Feature, now that the feature rule exists, to guard against regressions.


_Auto-resolved by the [GitHub Copilot app](https://github.com/github/app)._

File: src/EncDotNet.S100.Datasets.S57/S57ToS101Translator.cs [hunk 17]
```diff
@@ -451,6 +1205,444 @@ private void AppendInformationInstance(
...
+            // the sectorLimit subtree (itself [0..1]) is only emitted when both
+            // bearings are present; a lone bearing omits the whole subtree.
+            if (sectorBearingOne is not null && sectorBearingTwo is not null)
+            {
+                builder.Add(new S101Attribute(GetOrAssignAttributeCode(S101AttrSectorLimit), 1, string.Empty));
+                builder.Add(new S101Attribute(GetOrAssignAttributeCode(S101AttrSectorLimitOne), 1, string.Empty));
+                builder.Add(new S101Attribute(GetOrAssignAttributeCode(S101AttrSectorBearing), 1, sectorBearingOne));
+                builder.Add(new S101Attribute(GetOrAssignAttributeCode(S101AttrSectorLimitTwo), 1, string.Empty));
+                builder.Add(new S101Attribute(GetOrAssignAttributeCode(S101AttrSectorBearing), 1, sectorBearingTwo));
+            }
```
[the remark this answers is not in the document]
Comment (philliphoff) [reply src/EncDotNet.S100.Datasets.S57/S57ToS101Translator.cs:1678]: Accepted. When only one of SECTR1/SECTR2 is present the surviving bearing is now recorded via RecordRuleDroppedAttribute (SECTR1 or SECTR2), so corpus audits see the data loss, matching the PERSTA/PEREND and SURSTA/SUREND handling. Extended the lone-bearing test to assert the drop is recorded.


_Auto-resolved by the [GitHub Copilot app](https://github.com/github/app)._

File: src/EncDotNet.S100.Datasets.S57/S57ToS101Translator.cs [hunk 18]
```diff
@@ -425,6 +995,190 @@ private IReadOnlyList<S101Attribute> TranslateAttributes(
...
+                {
+                    AppendSectorCharacteristicsInstance(
+                        builder, sectrLitchr, sectrSiggrp, sectrSigper, sectrSigseq,
+                        sectrColour, sectrLitvis, sectrValnmr, sectrSectr1, sectrSectr2);
+                }
+                else
+                {
+                    _diagnostics?.RecordDroppedEnumValue(S101AttrLightCharacteristic, sectrLitchr);
+                }
+            }
```
[the remark this answers is not in the document]
Comment (philliphoff) [reply src/EncDotNet.S100.Datasets.S57/S57ToS101Translator.cs]: Accepted. When no sectorCharacteristics instance is emitted (LITCHR missing or FC-rejected), the diverted sector attributes (COLOUR/LITVIS/VALNMR/SECTR1/SECTR2/SIGGRP/SIGPER/SIGSEQ) are now recorded via RecordRuleDroppedAttribute (and an FC-rejected LITCHR via RecordDroppedEnumValue), so corpus audits see the loss. Added a test for the LITCHR-missing case.


_Auto-resolved by the [GitHub Copilot app](https://github.com/github/app)._

File: src/EncDotNet.S100.Datasets.S57/S57ToS101Translator.cs [hunk 19]
```diff
@@ -389,29 +773,215 @@ private IReadOnlyList<S101Attribute> TranslateAttributes(
...
+                        {
+                            _diagnostics?.RecordRuleDroppedAttribute(attl);
+                            continue;
+                        }
+
+                        if (!_allowedEnumValues.IsAllowed(sub.S101Code, sub.Value))
+                        {
+                            _diagnostics?.RecordDroppedEnumValue(sub.S101Code, sub.Value);
+                            continue;
+                        }
```
[the remark this answers is not in the document]
Comment (philliphoff) [reply src/EncDotNet.S100.Datasets.S57/S57ToS101Translator.cs]: Good catch. Split now happens independently of FC enforcement: when `_allowedEnumValues` is available I use its authoritative `IsEnumerated` flag, and when enforcement is disabled I fall back to a structural integer-list discriminator (`IsIntegerList`) so list enums like `COLOUR`/`CATLND` "1,3" are still split while free text such as `OBJNAM` "Smith, Jones and Co." is left intact. `IsAllowed` filtering is skipped in the disabled mode. Added regression tests for both the split-when-disabled and text-not-split cases (commit 3bd2011).


_Auto-resolved by the [GitHub Copilot app](https://github.com/github/app)._
